host-controller/src/main/java/org/jboss/as/host/controller/operations/StartServersHandler.java

```diff
@@ -95,27 +96,47 @@ public void execute(OperationContext context, ModelNode operation) throws Operat
+
+                    if (blockUntilStopped) {
+                        reconnectAndWaitStoppingServers(servers, domainModel, true);
+                    }
+
                     if (hostControllerEnvironment.isRestart() || runningModeControl.getRestartMode() == RestartMode.HC_ONLY){
-                        restartedHcStartOrReconnectServers(servers, domainModel, context);
-                        runningModeControl.setRestartMode(RestartMode.SERVERS);
+                        restartedHcStartOrReconnectServers(servers, domainModel, context, restartRequired);
                     } else {
-                        cleanStartServers(servers, domainModel, context);
+                        cleanStartServers(servers, domainModel, context, restartRequired);
                     }
                 }
+                //Reset runningModeControl values to default for the values that cannot be reset at ProcessReloadHandler.ReloadContext.doReloado
+                runningModeControl.setRestartMode(RestartMode.SERVERS);
+                runningModeControl.setBlockUntilStopped(true);
                 context.completeStep(OperationContext.RollbackHandler.NOOP_ROLLBACK_HANDLER);
             }
         }, OperationContext.Stage.RUNTIME);
     }
 
-    private void cleanStartServers(final ModelNode servers, final ModelNode domainModel, OperationContext context) throws OperationFailedException {
+    private void reconnectAndWaitStoppingServers(final ModelNode servers, final ModelNode domainModel, boolean blockUntilStopped){
+        Map<String, ProcessInfo> processInfos = serverInventory.determineRunningProcesses();
+        for(final String serverName : servers.keys()) {
+            ProcessInfo info = processInfos.get(serverInventory.getServerProcessName(serverName));
+            if (info != null){
+                serverInventory.reconnectServer(serverName, domainModel, info.getAuthKey(), info.isRunning(), info.isStopping(), blockUntilStopped, false, false);
+            }
+        }
+    }
+
+    private void cleanStartServers(final ModelNode servers, final ModelNode domainModel, OperationContext context, boolean restartRequired) throws OperationFailedException {
         Map<String, ProcessInfo> processInfos = serverInventory.determineRunningProcesses();
         for(final Property serverProp : servers.asPropertyList()) {
             String serverName = serverProp.getName();
             if (ServerConfigResourceDefinition.AUTO_START.resolveModelAttribute(context, serverProp.getValue()).asBoolean(true)) {
                 ProcessInfo info = processInfos.get(serverInventory.getServerProcessName(serverName));
                 if ( info != null ){
-                    serverInventory.reconnectServer(serverName, domainModel, info.getAuthKey(), info.isRunning(), info.isStopping());
+                    serverInventory.reconnectServer(serverName, domainModel, info.getAuthKey(), info.isRunning(), info.isStopping(), false, restartRequired, START_BLOCKING);
```

Comment:
I found an issue here, we can get duplicate registration error because we are only reconnecting here servers which have the auto-start set. For example, if we are reloading with server-three started (auto-start=false), we will get a duplicate registration executing `:start-servers` after reloading, because server-three was not reconnected in the HC reload, I will fix it.